### DISCUSSION TITLE
Sample cweb API output

### DISCUSSION THREAD

**Discussion by zmajeed** [gogotrango/cwebdraftapi#10] [General]
I'm fleshing out a C++ callbacks API for the cweb parser. The working cweb grammar is at https://github.com/gogotrango/cwebdraftapi/blob/main/grammar/cweb.grammar.bison.y

There will be a class say `CWeb` that will roughly take a bunch of callbacks in a struct and run the callbacks as it parses input. The library will be threadsafe in the sense that multiple `CWeb` objects could exist and be used concurrently though a single `CWeb` instance could not be shared across threads.

The callbacks mostly correspond to the symbols in the grammar.

So there is a `document_begin` callback for the root `cweb` symbol in the grammar. And a `document_end` for when `cweb` successfully reduces.

Similarly there is a `tex_begin` callback when a `tex` part begins - and a `tex_end` callback when the part ends

Each `@` control code has a callback. Some of these callbacks come in begin and end pairs, e.g. `c_named_section_begin` and `c_named_section_end` while others like `cweb_comment` contain all the data for the code

I've added a rough sample of the data the callbacks will offer in https://github.com/gogotrango/cwebdraftapi/tree/main/doc/cwebapi_design/cwebapi_sampleoutput.20210621.txt

The cweb test input files are in https://github.com/gogotrango/cwebdraftapi/tree/main/testfiles

Here are some snippets from the file - each line is like one callback call - a line starts with the callback name - then there's a sequence of colon-separated parameter-value pairs, e.g. `file_begin` callback has a `file` parameter with a filename string - two pairs are separated by commas

Original source text will be available in virtually all callbacks. The location and extent of the text will be available as a `begin` line and column numbers and an `end` line and column

```
document_begin:
file_begin: file: "miles_span.w"
limbo_begin: loc.begin: 1 1, loc.end: 2 1
free_text: type: "tex", loc.begin: 1 1, loc.end: 2 1, text: "% This file is part of the Stanford GraphBase (c) Stanford University 1993
"
include_file: file: "boilerplate.w", quoted: false, trailingText: " %<< legal stuff: PLEASE READ IT BEFORE MAKING ANY CHANGES!", loc.begin: 2 1, loc.end: 3 1, text: " boilerplate.w %<< legal stuff: PLEASE READ IT BEFORE MAKING ANY CHANGES!
"
```
```
file_begin: file: "boilerplate.w"
limbo_begin: loc.begin: 1 1, loc.end: 39 1
free_text: type: "tex", loc.begin: 1 1, loc.end: 39 1, text: "% This material goes at the beginning of all Stanford GraphBase CWEB files
```
```
limbo_end: loc.begin: 39 1, loc.end: 39 1
file_end: file: "boilerplate.w"
```
```
tex_begin: loc.begin: 9 1, loc.end: 9 1
tex_section_begin: type: "starred", title: "@* Minimum spanning trees.", loc.begin: 9 1, loc.end: 11 1, text: "@* Minimum spanning trees."
free_text: type: "tex", loc.begin: 9 1, loc.end: 11 1, text: "
A classic paper by R. L. Graham and Pavol Hell about the history of
"
index_entry: entry: "Graham, Ronald Lewis", loc.begin: 11 1, loc.end: 11 23, text: "@^Graham, Ronald Lewis@>"
```

This is a preliminary sketch of the API to show the direction I'm going in. Much remains to hash out - error reporting, callbacks registration, cancellation.

Please do let me know what you think if you're at all interested in developing cweb applications or even just cweb in general.

**Comment by shreevatsa** [comment]
This looks like great progress, thanks for the update @zmajeed!

Hopefully others will have more substantive comments, but a couple of things, just as filler:

> The library will be threadsafe in the sense that multiple `CWeb` objects could exist and be used concurrently though a single `CWeb` instance could not be shared across threads.

FWIW I'm not sure there's standard terminology around this, but I've seen this called "thread-compatible" (e.g. [here](https://abseil.io/blog/20180531-regular-types)).

> The callbacks mostly correspond to the symbols in the grammar.

This is cool! Having the original source text available in the callbacks also sounds very useful.

I haven't thought about it myself, but it seems it may be useful to keep a few applications in mind while thinking about this: how could this parsing API be used to implement, say:

1. CTANGLE,
2. CWEAVE,
3. a CWEB to HTML (say) translator

Obviously parsing isn't everything and most of these will require writing some higher-level code (e.g. imagine writing a CWEB to Rust translator!), but at least at first glance the parsing API described above seems like it will be useful and I can't see any obvious limitations that would get in the way.

**Reply by zmajeed** [reply]
Great advice for developing any API! Have some applications in mind - better yet have sample code use cases throughout to see how well the API works for specific tasks. Luckily I started with a slightly tweaked ctangle in mind - that's been my touchstone so far

Thanks for the abseil link on thread safety - I'd rather spell out more plainly what can be done with the API in multiple threads than use terminology that would require much explanation. I think jargon would be useful if different categories of thread support were to be distinguished.

Having source text readily available is so useful for developing and debugging the API - yet I can see some applications may not be interested in retaining that level of fidelity - so I might make location and source text access optional - maybe through a config setting

I find making the API roughly match the grammar lets one look at the grammar to understand how to use the API on a cweb document. If you're only interested in the C parts of a cweb you can kind of scope the structure of a C part in the grammar and figure out the callbacks you'll need. Instead it can be quite overwhelming if you just look a list of dozens of callbacks to know how to put them together

**Comment by texdraft** [comment]
Will it be possible to get the definitions/replacement text of a section?

Also, the sample output seems to be dropping one character from section names.

**Reply by zmajeed** [reply]
That's a pretty essential need with cweb isn't it! I don't think it will be built into the parsing API. But I can imagine providing conveniences like canned sets of callbacks for simple applications, e.g. echo or `cat` that simply echo source text. Similarly a class for named sections could be offered and used in another set of canned callbacks for automatic section management.

```c++
class NamedSection {
  string name ;
  string replacement;
  vector<Location> xref;
  NamedSection& add(string replacement);
  NamedSection& cite(Location loc);
};

class NamedSections {
  NamedSection* get(string prefix); 
  NamedSection& add(NamedSection section);
};
```

So now this is really `cweblib` - more than just a parsing API - a set of classes broken out of `cweave` and `ctangle` - I think it's great because it makes more of Knuth's work accessible as smaller components. His `name_info` data structure for section storage is actually pretty cool and I can see it benefitting many applications.

Thanks for noticing the section name truncation - there's a lot to get just right before the API is ready for public testing and every little bit of review helps

**Comment by zmajeed** [comment]
Couple thoughts in general for my notes

I think multiple callbacks for the same event would be nice - so a simple struct of callbacks may not be the right interface

Also user data is important for any parsing API - so there will be a way to set user data and access it in any callback

**Comment by zmajeed** [comment]
Updated grammar and sample output from API

https://github.com/gogotrango/cwebdraftapi/blob/main/grammar/cweb.grammar.bison.y
https://github.com/gogotrango/cwebdraftapi/blob/main/doc/cwebapi_design/cwebapi_sampleoutput.20210629.txt

Recognize C strings and C++ raw strings - better inner C support - removed lot of clutter from sample output

Still some drift in event locations toward the end of the document

Changed some callback/event names like named section reference is now just `section_name` - added an `isPrefix` parameter to indicate if section name has trailing ellipsis
